File: src/sub/dist.c
#include <math.h>
#include <stdio.h>

extern double olat, olon;
extern double rlatc, rad;
extern double aa, bb, sint, cost;

void dist(double xlat, double xlon, double *xkm, double *ykm) {
    fprintf(stderr ,"DEBUG dist inputs:\n");
    fprintf(stderr ,"  xlat=%.10f xlon=%.10f\n", xlat, xlon);

    fprintf(stderr ,"DEBUG globals:\n");
    fprintf(stderr ,"  olat=%.10f olon=%.10f\n", olat, olon);
    fprintf(stderr ,"  rlatc=%.15f rad=%.15f\n", rlatc, rad);
    fprintf(stderr ,"  aa=%.10f bb=%.10f\n", aa, bb);
    fprintf(stderr ,"  cost=%.10f sint=%.10f\n", cost, sint);
    double q, yp, xx;
    double lat1, lat2, lat3;

    if (!xkm || !ykm) {
        return;
    }

    q = 60.0 * xlat - olat;
    yp = q + olat;
    lat1 = atan(rlatc * tan(rad * yp / 60.0));
    lat2 = atan(rlatc * tan(rad * olat / 60.0));
    lat3 = (lat2 + lat1) / 2.0;

    xx = 60.0 * xlon - olon;
    q = q * aa;
    xx = xx * bb * cos(lat3);

    yp = cost * q - sint * xx;
    xx = cost * xx + sint * q;
    q = yp;

    *xkm = xx;
    *ykm = q;
    fprintf(stderr, "DEBUG intermediate variables:\n");
    fprintf(stderr, "  q=%.10f yp=%.10f xx=%.10f\n", q, yp, xx);
    fprintf(stderr, "  lat1=%.15f lat2=%.15f lat3=%.15f\n", lat1, lat2, lat3);
}

File: src_fortran/sub/dist.f
c
c
c
c
c
c
      subroutine DIST(xlat,xlon,xkm,ykm)   ! 7. Jan. 1988
c
      implicit none
      real xlat,xlon,xkm,ykm
c      
      real olat,olon,aa,bb,bc,sint,cost,rotate
      real q,yp,xx
      integer icoordsystem
      common/GEO_COORSYSTEM/olat,olon,
     &                      rearth,ellip,rlatc,rad,
     &                      aa,bb,bc,sint,cost,rotate, icoordsystem
      double precision rearth,ellip,rlatc,rad
c
      DOUBLE PRECISION LAT1,LAT2,LAT3
c     conversion of latitude and longitude to kilometers relative
c     to center of coordinates
c
      write(*,*) 'DEBUG dist inputs:'
      write(*,*) '  xlat=', xlat, ' xlon=', xlon

      write(*,*) 'DEBUG globals:'
      write(*,*) '  olat=', olat, ' olon=', olon
      write(*,*) '  rlatc=', rlatc, ' rad=', rad
      write(*,*) '  aa=', aa, ' bb=', bb
      write(*,*) '  cost=', cost, ' sint=', sint
      q=60*xlat-olat
      yp=q+olat
      lat1=datan(rlatc*dtan(RAD*yp/60.0))
      lat2=datan(rlatc*dtan(RAD*OLAT/60.0))
      LAT3=(LAT2+LAT1)/2.
      xx=60*xlon-olon
      q=q*aa
      xx=xx*bb*dcos(LAT3)
cc** rotate coordinate system clockwise
      yp=cost*q-sint*xx
      xx=cost*xx+sint*q
      q=yp
c
      xkm=xx
      ykm=q
      write(*,*) 'DEBUG intermediate variables:'
      write(*,*) '  q=', q, ' yp=', yp, ' xx=', xx
      write(*,*) '  lat1=', lat1, ' lat2=', lat2, ' lat3=', lat3
c
      return
      end ! of subr. dist
